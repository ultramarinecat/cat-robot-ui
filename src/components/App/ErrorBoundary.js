//
// Main error boundary, displays fallback ui indicating that an error has occurred
//

import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { withTranslation } from 'react-i18next';

import styles from './ErrorBoundary.module.scss';

export class ErrorBoundary extends Component {
  state = { hasError: false };

  static propTypes = {
    t: PropTypes.func.isRequired
  };

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  render() {
    // if error, display fallback ui
    if (this.state.hasError) {
      const { t } = this.props;

      return (
        <div className={styles.main} data-testid="error-fallback-ui">
          {t('error')}
        </div>
      );
    }

    return this.props.children;
  }
}

export default withTranslation()(ErrorBoundary);
