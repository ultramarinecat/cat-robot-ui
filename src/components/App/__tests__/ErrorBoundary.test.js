import React from 'react';
import { render } from 'react-testing-library';

import { ErrorBoundary } from '../ErrorBoundary.js';

const FALLBACK_UI = 'error-fallback-ui';
const CAT = 'cat';
const DOG = 'dog';

let consoleError;

const Cat = () => <div>{CAT}</div>;
const Dog = () => {
  throw new Error('dog!');
  return <div>{DOG}</div>; // eslint-disable-line no-unreachable
};

describe('ErrorBoundary', () => {
  beforeAll(() => {
    consoleError = console.error;
    console.error = jest.fn();
  });

  afterAll(() => {
    console.error = consoleError;
  });

  it('should render children when an error is not thrown by a descendant component', () => {
    const { container, queryByText, queryByTestId } = render(
      <ErrorBoundary t={key => key}>
        <Cat />
      </ErrorBoundary>
    );

    expect(queryByText(CAT)).toBeInTheDocument();
    expect(queryByTestId(FALLBACK_UI)).toBeNull();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display fallback ui when an error is thrown by a descendant component', () => {
    const { container, queryByText, queryByTestId } = render(
      <ErrorBoundary t={key => key}>
        <Dog />
      </ErrorBoundary>
    );

    expect(queryByTestId(FALLBACK_UI)).toBeInTheDocument();
    expect(queryByText(DOG)).toBeNull();

    expect(container.firstChild).toMatchSnapshot();
  });
});
