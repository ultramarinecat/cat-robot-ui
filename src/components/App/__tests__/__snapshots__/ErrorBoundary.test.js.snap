// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should display fallback ui when an error is thrown by a descendant component 1`] = `
<div
  class="main"
  data-testid="error-fallback-ui"
>
  error
</div>
`;

exports[`ErrorBoundary should render children when an error is not thrown by a descendant component 1`] = `
<div>
  cat
</div>
`;
